Add function to obtain relay ed25519

juga0 opened the pull request:
Not sure this is the best way to obtain it. Please review.

teor2345 commented on sbws/core/scanner.py:
```diff
@@ -131,7 +132,25 @@ def measure_bandwidth_to_server(session, conf, dest, content_length):
+    if relay.identifier is None:
+        try:
+            log.debug('Getting microdescriptor to obtain ed25519 identity.')
+            mdesc = controller.get_microdescriptor(relay.fingerprint)
```
This code only works for clients with "UseMicrodescriptors 1" (or auto), which is the default.

But many torflow clients have "UseMicrodescriptors 0", and we want to support clients that have server descriptors. So if getting the microdescriptor fails, please fall back to getting the server descriptor:

https://stem.torproject.org/api/control.html#stem.control.Controller.get_server_descriptor

teor2345 replied:
You'll want the ed25519_master_key of the RelayDescriptor:
https://stem.torproject.org/api/descriptor/server_descriptor.html#stem.descriptor.server_descriptor.RelayDescriptor

teor2345 commented on sbws/core/scanner.py:
```diff
@@ -131,7 +132,25 @@ def measure_bandwidth_to_server(session, conf, dest, content_length):
+
+
 def measure_relay(args, conf, destinations, cb, rl, relay):
+    obtain_ed25519(cb.controller, relay)
```
Please also try to obtain the ed25519 id when the relay object is created.

juga0 replied:
I wanted to do like that, but Relay object is only created at the end, when dumping results, as part of Result object: https://github.com/pastly/simple-bw-scanner/blob/master/sbws/lib/resultdump.py#L150
The first place where 1st is obtained descriptors (not relays) is in https://github.com/pastly/simple-bw-scanner/blob/master/sbws/lib/relaylist.py#L134
Without refactoring code, 2 possible solutions:
- obtain ed25519 when descriptors are obtained, and modify descriptor object
- obtain ed25519 in Result and/or Relay object
Which one seems better?

juga0 replied:
or other ideas where to do it? (@pastly)

teor2345 commented on sbws/core/scanner.py:
```diff
@@ -152,6 +171,7 @@ def measure_relay(args, conf, destinations, cb, rl, relay):
         # TODO: Return ResultError of some sort
         return None
     exit = rng.choice(exits)
+    obtain_ed25519(cb.controller, exit)
```
Please also try to obtain the ed25519 id when the relay object is created.

teor2345 commented on sbws/core/scanner.py:
```diff
@@ -131,7 +132,25 @@ def measure_bandwidth_to_server(session, conf, dest, content_length):
+            mdesc = controller.get_microdescriptor(relay.fingerprint)
+            if 'ed25519' in mdesc.identifiers.keys():
+                relay.identifier_type = 'ed25519'
+                relay.identifier = mdesc.identifiers['ed25519']
```
Maybe the function should strip any trailing = signs here.

juga0 commented:
Regarding the tests, this was WIP and there isn't any test for scanner.py, where i created it.
I'm not sure which is the best way right now to test this, since we aren't using chutney (testnets instead) and it would need to launch or connect to Tor (don't think i can mock it).
Unless there're other better ideas, i think i'd launch tor in conftest.py so it can be reused for other tests.

juga0 commented:
Instead of obtaining the key in ``scanner``, now  it's obtained when creating the ``RelayList``.
I added a test that needs to start Tor. While not using ``chutney``, i don't know a better way to do this.
I'm not modifying the Bandwidth List file lines, since that part of the code will change after merging #175 and #176. I'd modify that once those PRs are merged.

juga0 commented:
Travis fail due tor not being installed. I could add that to Travis, though i'd also add ``chutney``

pastly commented:
We've decided this doesn't matter for minimal viable product (1.0) so I'm removing that as a milestone.

juga0 commented:
Rebased to master.

pastly commented on sbws/core/generate.py:
```diff
@@ -24,6 +24,7 @@ def __init__(self, fp, bw, nick, rtts, last_time):
         self.time = unixts_to_isodt_str(last_time)
+        self.ed25519 = ed25519
 
     def __str__(self):
```
The point of getting the ed25519 identity is to include it on the line, right? So if it is not None, we should include it here, right?

juga0 replied:
yes, the point it's to print it in the bw line.
i think i put a comment somewhere that i was not adding that until #160 was merged, cause #160 was rewriting all that code.
i should rebase this to master again and print it in the bw line

pastly commented on sbws/lib/relaylist.py:
```diff
@@ -11,6 +13,18 @@
 log = logging.getLogger(__name__)
 
 
+class RelayNS(RouterStatusEntryV3):
```
I think instead of inheriting RouterStatusEntryV3, we should

- not inherit anything
- rename to Relay (since we're going to pull info from more than the network status consensus)
- init with just a fingerprint
- use the fingerprint to ask stem for relay information (like with `get_network_status` and/or `get_server_descriptor` and/or `get_microdescriptor`) store what we get
- define `@property`s for all the things we need like exit_policy and fingerprint and nickname and flags and its ed25519 certificate and ...

juga0 replied:
sounds good, maybe we should rename some "relay" vars that won't be Relay objs after these changes (not sure if/which now)

pastly commented on sbws/lib/relaylist.py:
```diff
@@ -68,6 +82,40 @@ def measured(self):
         # return [r for r in relays if r.measured is not None]
         return [r for r in relays if not r.is_unmeasured]
 
+    def relay_ed25519_master_key(self, ns):
```
I think this function doesn't belong in RelayList (it does something for a single relay, not something related to the RelayList as a whole); if the relay class is implemented as I envision, this function will aide the creation of `Relay.__init__` and then disappear.

juga0 replied:
i agree, i put it there cause is where currently the info for bw lines is coming from.

pastly commented:
I'm inclined to delete the tests that are partly causing the merge conflict.

juga0 commented:
> i think i put a comment somewhere that i was not adding that until #160 was merged, cause #160 was rewriting all that code.
> i should rebase this to master again and print it in the bw line

@pastly you forgot to rebase to master first. 
I've done it in https://github.com/juga0/simple-bw-scanner/tree/obtain_ed25519_rebased_master.
I was going to cherry-pick your commits and modify as needed, but maybe you prefer to do itf?

juga0 commented:
I tried rebasing first on top of your changes, but it was too many conflicts

juga0 commented:
ok, i'm cherry-picking your commits, resolving conflicts and solving new thinks that broke. Will push in the next hours

pastly commented:
Thanks @juga0. You can delete tests that are hard to fix.

juga0 commented:
Rebased also in this branch, cherry-picked your commits (solving merge conflicts) and fixed some tests.
The last two commits contain FIXMEs about tests not working because of Result changes.
Would you check if easy to fix for you?, since you created/modified that class.
Otherwise let me know and i fix them.
Thanks!

juga0 commented:
I'm fixing them

juga0 commented:
oh!, i didn't say it was ready to merge yet, i'm still fixing things...

pastly commented:
Crap I need to communicate better with you.

Well I'll merge those too when you're ready.

juga0 commented:
ok, thx

juga0 commented:
Last commit was just s/ed25519_master_key/master_key_ed25519 to unify the terms. Except for ``_from_desc('ed25519_master_key')``, as stem uses that term instead.
If you are fine with last commit, is ready to merge.
I don't see last commits here, maybe because it has been merged, but they're in my branch.

pastly commented:
Grabbed your docstring commit and the variable renaming commit. Thanks for all your work on this issue.
